Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -1797,16 +1772,48 @@ extension SourceKitServer {
       if URL(fileURLWithPath: occurrence.location.path).pathExtension == "swiftinterface" {
         // If the location is in `.swiftinterface` file, use moduleName to return textual interface.
         return try await self.definitionInInterface(
-          req,
           moduleName: occurrence.location.moduleName,
           symbolUSR: occurrence.symbol.usr,
+          originatorUri: uri,
           languageService: languageService
         )
       }
       return indexToLSPLocation(occurrence.location)
     }
   }
 
+  /// Returns the result of a `DefinitionRequest` by running a `SymbolInfoRequest`, inspecting
+  /// its result and doing index lookups, if necessary.
+  ///
+  /// In contrast to `definition`, this does not fall back to sending a `DefinitionRequest` to the
+  /// toolchain language server.
+  private func indexBasedDefinition(
+    _ req: DefinitionRequest,
+    workspace: Workspace,
+    languageService: ToolchainLanguageServer
+  ) async throws -> [Location] {
+    let symbols = try await languageService.symbolInfo(
+      SymbolInfoRequest(
+        textDocument: req.textDocument,
+        position: req.position
+      )
+    )
+    if symbols.allSatisfy({
+      $0.bestLocalDeclaration != nil && !($0.isDynamic ?? true)
+        && $0.usr?.hasPrefix("s:") ?? false /* Swift symbols have USRs starting with s: */
+    }) {
+      // If we have a known non-dynamic symbol within Swift, we don't need to do an index lookup.
+      // For non-Swift symbols, we need to perform an index lookup because the best local declaration will point to
+      // a header file but jump-to-definition should prefer the implementation (there's the declaration request to jump
+      // to the function's declaration).
+      return symbols.compactMap { $0.bestLocalDeclaration }
+    }
+
+    return try await symbols.asyncMap {
+      try await self.definitionLocations(for: $0, in: req.textDocument.uri, languageService: languageService)
```

Comment:
Rather than do the `allSatisfy` above and then mapping all here, could we just map each as required? Eg. there's no reason to run `definitionLocations` on non-dynamic Swift symbols.

Reply:
Good idea 👍🏽